addrman: change internal id counting to int64_t

mzumsande opened the pull request:
With `nIdCount` being incremented for each addr received, an attacker could cause an overflow in the past, see https://bitcoincore.org/en/2024/07/31/disclose-addrman-int-overflow/
Even though that attack was made infeasible indirectly by addr rate-limiting (PR #22387), to be on the safe side and prevent any regressions change the `nId`s used internally to `int64_t`.
This is being done by first introducing a user-defined type for `nId`s in the first commit, and then updating it to `int64_t` (thanks sipa for help with this!).

Note that `nId` is only used internally, it is not part of the serialization, so `peers.dat` should not be affected by this.

I assume that the only reason this was not done in the past is to not draw attention to this previously undisclosed issue.

mzumsande committed:
addrman, refactor: introduce user-defined type for internal nId

This makes it easier to track which spots refer to an nId
(as opposed to, for example, bucket index etc. which also use int)

Co-authored-by: Pieter Wuille <pieter@wuille.net>

hebasto commented on src/addrman_impl.h:
```diff
@@ -178,8 +178,10 @@ class AddrManImpl
+    //! last used nId.
+    //! This used to be an int, making it feasible for attackers to cause an overflow,
+    //! see https://bitcoincore.org/en/2024/07/31/disclose-addrman-int-overflow/
+    uint64_t nIdCount GUARDED_BY(cs){0};
```
Amend the `pnId` parameter type in `AddrManImpl::Create` respectively?

mzumsande replied:
true. It also needs more changes (`mapInfo`). I wil put te PR to draft until I have addressed these.

mzumsande replied:
Should be addressed now.

mzumsande committed:
addrman: change nid_type from int to int64_t

With nId being incremented for each addr received,
an attacker could cause an overflow in the past.
(https://bitcoincore.org/en/2024/07/31/disclose-addrman-int-overflow/)
Even though that attack was made infeasible by
rate-limiting (PR #22387), to be on the safe side change the
type to an int64_t.

dergoegge commented:
Concept ACK

brunoerg commented:
Concept ACK

mzumsande commented:
Updated the PR to have a refactor-only commit that introduces a user-defined type for all spots that use `nId` as an int, and then change that to `int64_t` in the second commit.

BrandonOdiwuor reviewed (commented):
Concept ACK

hebasto commented on src/addrman_impl.h:
```diff
@@ -32,6 +32,13 @@ static constexpr int ADDRMAN_NEW_BUCKET_COUNT{1 << ADDRMAN_NEW_BUCKET_COUNT_LOG2
+ * This used to be int, making it feasible for attackers to cause an overflow,
+ * see https://bitcoincore.org/en/2024/07/31/disclose-addrman-int-overflow/
+ */
+using nid_type = int64_t;
```
[Good](https://www.aristeia.com/Papers/C++ReportColumns/sep95.pdf) to see that a signed type is used, rather than an unsigned one in the initial version of this PR.

mzumsande replied:
interesting pdf, thanks! In this particular case there was no choice because `nId`s in `vvNew` and `vvTried` are [initialised](https://github.com/bitcoin/bitcoin/blob/2aff9a36c352640a263e8b5de469710f7e80eb54/src/addrman.cpp#L109-L125) to `-1`.

hebasto reviewed (commented):
Concept ACK.

mzumsande commented on src/addrman.cpp:
```diff
@@ -188,7 +188,7 @@ void AddrManImpl::Serialize(Stream& s_) const
     int nUBuckets = ADDRMAN_NEW_BUCKET_COUNT ^ (1 << 30);
     s << nUBuckets;
-    std::unordered_map<int, int> mapUnkIds;
+    std::unordered_map<nid_type, int> mapUnkIds;
```
Hmm, can you explain the "morally" part a bit more?

`mapUnkIds` maps `nid_type`  to an `int`, let's call that counter `ser_count`. `ser_count` is part of the serialization and is bounded by the limited slots of the addrman tables - even if the internal id would overflow an `int`, `ser_count` wouldn't be affected by that.

During deserialization, `ser_count` is read into `int entry_index` (which must be `int`), and then makes it into `bucket_entries`  and later into `vvNew`. But I'm unsure when is the best step to convert it into an `int64_t` -  why is it better to already do that in the step `entry_index` -> `bucket_entries` rather than the way it is now?

stratospher replied:
I'm fine with the version on the branch now because `Serialise()` and `Unserialise()` are symmetrical.  `nIndex` is [written as int](https://github.com/bitcoin/bitcoin/blob/51f7668d31e2624e41c7ce77fe33162802808f3f/src/addrman.cpp#L220) into peers.dat in `Serialise()` and [read as int](https://github.com/bitcoin/bitcoin/blob/51f7668d31e2624e41c7ce77fe33162802808f3f/src/addrman.cpp#L325) from peers.dat in `Unserialise()`.

stratospher reviewed (commented):
Concept ACK. if you retouch, there's also: https://github.com/bitcoin/bitcoin/blob/c2d15d993ef06d97d4c117012bda6efa3dcbac45/src/test/fuzz/addrman.cpp#L189

mzumsande commented:
> Concept ACK. if you retouch, there's also:

Good catch, fixed!

tdb3 reviewed (commented):
Concept ACK.
Briefly reviewed, and will return to review in more detail.  Love the approach to introduce `nid_type`.  Makes things safer moving forward.

Christewart commented:
Concept ACK

So presumably test cases are in in #22387 to directly test the vuln? I had a hard time deciphering what is meant to test rate limiting logic in #22387 and what test cases were for preventing a regression in the future.

stratospher commented:
> I had a hard time deciphering what is meant to test the new rate limiting logic in https://github.com/bitcoin/bitcoin/pull/22387 and what test cases were intended to prevent a regression for being introduced in the future that re-introduces this vulnerability.

Main thing is to make sure that `nId` values are distinct. #22387 limits the `nId` which the addrman can possibly construct([functional tests which test this behaviour](https://github.com/bitcoin/bitcoin/blob/ee367170cb2acf82b6ff8e0ccdbc1cce09730662/test/functional/p2p_addr_relay.py#L352)). This PR makes the `nId` space so large (64 bits) that the values are distinct. so even in the absence of the rate limiting mechanism, with this PR the attacker would need to insert > 2**64 addresses to make the node crash.

(you could change the type of `nIdCount` to `int8_t` to see the undesirable effects of duplicate `nId` in action!)


I looked into the assertion crash since I was confused how it happened. Sharing details in case it's helpful to anyone.

32 bit `nIdCount` could overflow and wrap around if addr messages were spammed (`nIdCount` would go from 0 to `INT32_MAX`, then become negative and go from `INT32_MIN` to 0 and loop again). So `nId`'s aren't distinct anymore. [`Create()`](https://github.com/bitcoin/bitcoin/blob/ee367170cb2acf82b6ff8e0ccdbc1cce09730662/src/addrman.cpp#L416) is written with an assumption that every `nId` it receives is distinct. [This line](https://github.com/bitcoin/bitcoin/blob/99ecb9a630e68f7fa5a449065020292790e3e7fb/src/addrman.cpp#L421) ends up overwriting the `nId`-> `AddrInfo` map without checking if it exists. This can lead to corruptions and assertion failures.

An example scenario:
- suppose `AddrInfo1` with `nId=36` is inserted into a bucket-position in `vvNew`(new table) and `mapInfo` now contains `nId=36` -> `AddrInfo1` mapping
- when `nId=36` happens again (because `nId` wraps around due to overflow) and another address `AddrInfo2` is inserted into the new table, there is no clean deletion of existing `nId` in `vvNew` /`mapAddr` and the `nId=36` in `vvNew` means `AddrInfo2` now.
- It's also possible that we don't want to insert `AddrInfo2` into new table and `nId=36` address can get [deleted later](https://github.com/bitcoin/bitcoin/blob/99ecb9a630e68f7fa5a449065020292790e3e7fb/src/addrman.cpp#L618).
- when the same bucket-position happens again, and we want to rewrite it's contents(`nId=36`) which meant the overwritten old address -> that info isn't present in `mapInfo` and `nRefCount = 0` (because it got deleted!)
- this will cause an assertion failure like the one below and the node crashes.
```
bitcoind: addrman.cpp:494: void AddrManImpl::ClearNew(int, int): Assertion `infoDelete.nRefCount > 0' failed.
Aborted (core dumped)
```

tdb3 commented on src/addrman_impl.h:
```diff
@@ -258,7 +265,7 @@ class AddrManImpl
      *
      *  @return  int The nid of the entry. If the addrman position is empty or not found, returns -1.
```
nit:  The return type in the Doxygen content could be adjusted.
```diff
-     *  @return  int The nid of the entry. If the addrman position is empty or not found, returns -1.
+     *  @return  nid_type The nid of the entry. If the addrman position is empty or not found, returns -1.
      * */
     nid_type GetEntry(bool use_tried, size_t bucket, size_t position) const EXCLUSIVE_LOCKS_REQUIRED(cs);
```

nit / thinking out loud:  Returning `-1` is a remnant of the previous type being a plain `int`.  To keep with the theme of using the user-defined `nid_type`, would it be worth creating a macro (or constexpr, or similar) to name the invalid case (e.g. `INVALID_NID = -1`)?  Not sure, as I'm not seeing a need to adjust `nid_type` to anything more complicated than `int64_t`, and `GetEntry()` in this case is just indicating empty/not found.

mzumsande replied:
Fixed the return type.

Yes, also not sure about `INVALID_NID` - I don't really see why we would want to change it from -1 to another value, but happy to add it if others like it... I'll keep this thread open for a while.

tdb3 reviewed (commented):
Approach ACK
Reviewed a bit more. Just nits.

mzumsande commented:
[1ded29d ](https://github.com/bitcoin/bitcoin/commit/1ded29db12c4efa0d575c63c50a07f002ff4c436)to [51f7668](https://github.com/bitcoin/bitcoin/commit/51f7668d31e2624e41c7ce77fe33162802808f3f):
Rebased and fixed doxygen comment.

> So presumably test cases are in https://github.com/bitcoin/bitcoin/pull/22387 to directly test the vuln?

No, there is no test case that directly tests the vulnerability. For that, we'd need to add approximately `2 * INT_MAX` addrs to addrman (see the explanation by @stratospher above) - which I think would be too slow to be practical for the test suite. However, one simple way to recreate the vulnerability (with this branch) is to locally change `nid_type` to `int8_t` - if you then run a node, it will crash quickly.

stratospher commented:
ACK 51f7668d31e2624e41c7ce77fe33162802808f3f. I think it's a good change to make the nId space large(64 bits) so that the nId values are distinct.

achow101 commented:
ACK 51f7668d31e2624e41c7ce77fe33162802808f3f

AlfazetOperations commented:
https://github.com/bitcoin-core/guix.sigs.git
